py/_io/terminalwriter.py

```diff
@@ -144,12 +144,30 @@ def __init__(self, file=None, stringio=False, encoding=None):
     def fullwidth(self):
         if hasattr(self, '_terminal_width'):
             return self._terminal_width
-        return get_terminal_width()
+        try:
+            width = self._cached_terminal_width
```

Comment:
Instead of implementing the entire cache logic here, how about setting up the cache invalidation support during __init__() and make each logical step in `fullwidth()` more well-defined?


```python

    def __init__(self, file=None, stringio=False, encoding=None):
        ...
        self._setup_terminal_width_cache()


    def _setup_terminal_width_cache():        
        if hasattr(signal, 'SIGWINCH'):
            signal.signal(signal.SIGWINCH, self._invalidate_cached_terminal_width)
            self._cached_terminal_width = None

    def _invalidate_cached_terminal_width(self):
        self._cached_terminal_width = None

     @property
      def fullwidth(self):
          # explicitly set by the user; return it directly 
          if hasattr(self, '_terminal_width'):
              return self._terminal_width
          # dynamic cache    
          if hasttr(self, '_cached_terminal_width'):
              if self._cached_terminal_width is None:
                  self._cached_terminal_width = get_terminal_width()
              return self._cached_terminal_width         
          # no dynamic cache: grab and return it directly
          return get_terminal_width()    

```

This makes the method much easier to understand IMHO. What do you think?